File: leetcode/17.电话号码的字母组合/17.电话号码的字母组合.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        # 组合，回溯算法
        mapping = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }
        need_list = []
        for s in digits:
            need_list.append(mapping.get(s, ''))
        need_str = "".join(need_list)
        result = []
        path = []

        def backtrack(path, need_str):
            if len(path) == len(need_list):
                result.append(''.join(path))
                return
            for s in need_list:
                path.append(s)
                backtrack(path, need_str)
                path.pop()

        backtrack(path, need_str)

File: leetcode/17.电话号码的字母组合/17.电话号码的字母组合.go
package main

import "strings"

func letterCombinations(digits string) []string {
	phone := map[byte][]string{
		'2': {"a", "b", "c"},
		'3': {"d", "e", "f"},
		'4': {"g", "h", "i"},
		'5': {"j", "k", "l"},
		'6': {"m", "n", "o"},
		'7': {"p", "q", "r", "s"},
		'8': {"t", "u", "v"},
		'9': {"w", "x", "y", "z"}}
	path := []string{}
	result := []string{}
	n := len(digits)
	if n == 0 {
		return result
	}
	var backtrack func(path []string, index int)
	backtrack = func(path []string, index int) {
		// 满足结束条件
		if len(path) == n {
			result = append(result, strings.Join(path, ""))
			return
		}
		// 循环选择
		digit := digits[index]
		for _, s := range phone[digit] {
			// 做选择
			path = append(path, s)
			backtrack(path, index+1)
			path = path[:len(path)-1]
		}
	}
	backtrack(path, 0)
	return result
}
